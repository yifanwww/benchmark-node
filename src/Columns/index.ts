export * from './ArgumentColumn';
export * from './ArgumentColumnHelper';
export * from './Column';
export * from './StatisticColumn';
export * from './StatisticColumnHelper';
export * from './TableColumn';
export * from './TableColumnHelper';
export * from './UnitType';
