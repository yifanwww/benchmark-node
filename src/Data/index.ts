export * from './ArgumentStore/ArgumentStore';
export * from './ArgumentStore/ArgumentStoreView';
export * from './Settings';
export * from './Statistics';
