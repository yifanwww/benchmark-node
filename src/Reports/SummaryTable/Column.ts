import type { Statistics } from '../../Data';

export type GetValue = (stats: Statistics) => string;

export class Column {
    protected declare readonly _columnName: string;

    protected declare _getValue: GetValue;

    constructor(columnName: string, getValue: GetValue) {
        this._columnName = columnName;

        this._getValue = getValue;
    }

    /**
     * Display column title.
     */
    get columnName(): string {
        return this._columnName;
    }

    get name() {
        return this._columnName;
    }

    /**
     * Formatted value of this column.
     */
    getValue(stats: Statistics): string {
        return this._getValue(stats);
    }
}
