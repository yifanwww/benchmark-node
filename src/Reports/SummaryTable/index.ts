import { ArgumentColumn, BaseColumn, Column, ColumnType, ParameterColumn } from '../../Columns';
import type { StatisticColumn } from '../../Columns';
import type { Statistics } from '../../Data';
import { TimeUnitHelper } from '../../Tools/TimeUnit';
import { UnitType } from '../../Tools/UnitType';
import { Report } from '../Report';
import type { ReportOptions } from '../Report';
import { ColumnAlign, createColumnInfo, Table } from '../Table';

export interface SummaryTableOptions extends ReportOptions {}

/**
 * @deprecated The temporary interface to describe data.
 */
export interface SummaryTableData {
    argLen: number;
    columns: readonly StatisticColumn[];
    paramNames: readonly string[];
    statsGroups: ReadonlyArray<readonly Statistics[]>;
}

export class SummaryTable extends Report<string> {
    private static readonly _fnColumn = new BaseColumn(ColumnType.FN, 'Function', (stats) => stats.name);

    generate(): this {
        // TODO

        return this;
    }

    // eslint-disable-next-line @typescript-eslint/naming-convention
    temporary_generate(data: SummaryTableData): this {
        const { argLen, columns, paramNames, statsGroups } = data;

        const table = new Table();
        table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, UnitType.ORIGIN));
        table.setHeader(0, SummaryTable._fnColumn.name);

        const paramColumns: ParameterColumn[] = [];
        const argColumns: ArgumentColumn[] = [];
        const statsColumns: StatisticColumn[] = [...columns];

        // FIXME: should not be `appending columns`

        for (let i = 0; i < paramNames.length; i++) {
            const col = new ParameterColumn(paramNames[i], i);
            paramColumns.push(col);
            table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, UnitType.ORIGIN));
            table.setHeader(i + 1, col.name);
        }

        for (let i = 0; i < argLen; i++) {
            const col = new ArgumentColumn(i);
            argColumns.push(col);
            table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, UnitType.ORIGIN));
            table.setHeader(i + paramColumns.length + 1, col.name);
        }

        // setup columns

        const start = table.columnCount;

        for (let i = 0; i < columns.length; i++) {
            const col = columns[i];
            table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, col.unitType));
            table.setHeader(start + i, col.name);
        }

        // setup statistics

        const helpers = [SummaryTable._fnColumn, ...paramColumns, ...argColumns, ...statsColumns];

        for (let groupId = 0; groupId < statsGroups.length; groupId++) {
            const group = statsGroups[groupId];
            for (let i = 0; i < group.length; i++) {
                for (let j = 0; j < helpers.length; j++) {
                    table.setCell([groupId, i], j, helpers[j].getData(group[i]) as never);
                }
            }
        }

        // generate table

        const report: string[] = [];

        table.chooseTimeUnit(table.findColumn(Column.Mean.name));
        report.push(table.render());

        if (this._description) {
            report.push('\nDescription:');
            const maxLen = statsColumns.reduce((prev, curr) => Math.max(prev, curr.name.length), 0);
            for (const column of statsColumns) {
                report.push(`- ${column.name.padEnd(maxLen)}: ${column.desc}`);
            }
            report.push(`- ${TimeUnitHelper.getFullDescription(table.timeUnit, maxLen)}`);
        }

        this._report = report.join('\n');

        return this;
    }
}
