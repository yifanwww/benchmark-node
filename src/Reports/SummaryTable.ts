import {
    ArgumentColumn,
    ArgumentColumnHelper,
    Column,
    StatisticColumn,
    StatisticColumnHelper,
    TableColumn,
    TableColumnHelper,
} from '../Columns';
import { Statistics } from '../Data';
import { ConsoleLogger } from '../Tools/ConsoleLogger';
import { TimeUnit, TimeUnitHelper } from '../Tools/TimeUnit';

export class SummaryTable {
    private static readonly fnNameColumn = new TableColumn('Function', (stats) => stats.name);
    private static readonly argColumn: ArgumentColumn[] = [];

    private declare readonly _fnNameColumnHelper;
    private declare readonly _argColumnHelpers: ArgumentColumnHelper[];
    private declare readonly _statsColumnHelpers: StatisticColumnHelper[];

    private declare readonly _statsArr: Statistics[];

    private declare _timeUnit: TimeUnit;

    public constructor() {
        this._fnNameColumnHelper = new TableColumnHelper(SummaryTable.fnNameColumn);
        this._argColumnHelpers = [];
        this._statsColumnHelpers = [];

        this._statsArr = [];

        this._timeUnit = TimeUnit.NS;
    }

    public addStatisticColumns(columns: StatisticColumn[]): void {
        for (const column of columns) {
            this._statsColumnHelpers.push(new StatisticColumnHelper(column));
        }
    }

    public addStats(statsArr: Statistics[]): void {
        this._statsArr.push(...statsArr);

        const maxLength = statsArr.reduce((prev, curr) => Math.max(prev, curr.args?.args.length ?? 0), 0);
        for (let i = this._argColumnHelpers.length; i < maxLength; i++) {
            if (i >= SummaryTable.argColumn.length) {
                SummaryTable.argColumn.push(new ArgumentColumn(i));
            }
            this._argColumnHelpers.push(new ArgumentColumnHelper(SummaryTable.argColumn[i]));
        }
    }

    private setFractionDigit(): void {
        let minTime: number;

        for (const helper of this._statsColumnHelpers) {
            if (helper.column.columnName === Column.Mean.columnName) {
                minTime = helper.findMinNumber(this._statsArr);
                break;
            }
        }

        this._timeUnit = TimeUnitHelper.chooseUnit(minTime!);
        for (const helper of this._statsColumnHelpers) {
            helper.timeUnit = this._timeUnit;
            helper.findFractionDigit(this._statsArr);
        }
    }

    public drawSummaryTable(): void {
        this.setFractionDigit();

        const helpers = [this._fnNameColumnHelper, ...this._argColumnHelpers, ...this._statsColumnHelpers];

        for (const helper of helpers) {
            helper.calculateMaxLen(this._statsArr);
        }

        const logger = ConsoleLogger.default;

        logger.writeLineStatistic(`| ${helpers.map((helper) => helper.drawColumnName()).join(' | ')} |`);
        logger.writeLineStatistic(`|-${helpers.map((helper) => helper.drawSperator()).join('-|-')}-|`);
        for (const stats of this._statsArr) {
            logger.writeLineStatistic(`| ${helpers.map((helper) => helper.draw(stats)).join(' | ')} |`);
        }

        logger.writeLine();
    }

    public writeDescription(): void {
        const logger = ConsoleLogger.default;

        logger.writeLineStatistic('Description:');
        const maxLen = this._statsColumnHelpers.reduce(
            (prev, curr) => Math.max(prev, curr.column.columnName.length),
            0,
        );
        for (const helper of this._statsColumnHelpers) {
            logger.writeLineStatistic(`- ${helper.column.columnName.padEnd(maxLen)}: ${helper.column.desc}`);
        }
        logger.writeLineStatistic(`- ${TimeUnitHelper.getFullDescription(this._timeUnit, maxLen)}`);
    }
}
