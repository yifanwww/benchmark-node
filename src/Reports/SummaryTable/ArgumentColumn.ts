import { Formatter } from '../../Tools/Formatter';

import { Column } from './Column';

export class ArgumentColumn extends Column {
    constructor(index: number) {
        super(`arg ${index}`, (stats) => {
            const args = stats.args?.args;

            if (!args || index >= args.length) {
                return '?';
            }

            const arg = args[index];
            const typeofArg = arg;

            if (
                typeofArg === 'bigint' ||
                typeofArg === 'boolean' ||
                typeofArg === 'number' ||
                typeofArg === 'symbol' ||
                typeofArg === 'undefined'
            ) {
                return String(args);
            }

            if (typeofArg === 'function') {
                return 'function';
            }

            // If argument is too long
            return Formatter.limitStringLength(String(arg));
        });
    }
}
