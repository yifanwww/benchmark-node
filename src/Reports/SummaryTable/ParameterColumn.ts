import { Formatter } from '../../Tools/Formatter';

import { Column } from './Column';

export class ParameterColumn extends Column {
    constructor(title: string, index: number) {
        super(title, (stats) => {
            const param = stats.params![index];

            const typeofParam = typeof param;

            if (
                typeofParam === 'bigint' ||
                typeofParam === 'boolean' ||
                typeofParam === 'number' ||
                typeofParam === 'symbol' ||
                typeofParam === 'undefined'
            ) {
                return String(param);
            }

            if (typeofParam === 'function') {
                return 'function';
            }

            // If parameter is too long
            return Formatter.limitStringLength(String(param));
        });
    }
}
