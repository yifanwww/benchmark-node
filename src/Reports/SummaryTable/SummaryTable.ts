import { StatisticIndicator } from '../../Indicators';
import type { IIndicator } from '../../Indicators';
import { RuntimeInfo } from '../../RuntimeInfo';
import { TimeUnitHelper } from '../../Tools/TimeUnit';
import type { TimeUnit } from '../../Tools/TimeUnit';
import { UnitType } from '../../Tools/UnitType';
import { Reporter } from '../Reporter';
import type { BenchmarkResult, ReporterOptions } from '../Reporter';
import { ColumnAlign, createColumnInfo, Table } from '../Table';

import { ArgumentColumn } from './ArgumentColumn';
import { Column } from './Column';
import { ParameterColumn } from './ParameterColumn';

export interface SummaryTableOptions extends ReporterOptions {}

export interface SummaryTableReport {
    description: string | undefined;
    runtime: string | undefined;
    table: string;
}

export class SummaryTable extends Reporter<SummaryTableReport> {
    private static readonly _fnColumn = new Column('Function', (stats) => stats.fnName);

    private _generateRuntimeInfo(): string {
        return [
            `BenchmarkNode v${RuntimeInfo.version}, ${RuntimeInfo.platform}`,
            RuntimeInfo.cpu.toString(),
            `Node.JS ${RuntimeInfo.versions.node} (V8 ${RuntimeInfo.versions.v8})`,
        ].join('\n');
    }

    private _generateDescription(indicators: readonly IIndicator[], timeUnit: TimeUnit): string {
        const maxLen = indicators.reduce((prev, curr) => Math.max(prev, curr.indicatorName.length), 0);

        const description = [
            'Description:',
            ...indicators.map((indicator) => `- ${indicator.indicatorName.padEnd(maxLen)}: ${indicator.legend}`),
            `- ${TimeUnitHelper.getFullDescription(timeUnit, maxLen)}`,
        ];

        return description.join('\n');
    }

    generate(result: BenchmarkResult): this {
        const { argLen, indicators, paramNames, statisticGroups } = result;

        const table = new Table();
        table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, UnitType.ORIGIN));
        table.setHeader(0, SummaryTable._fnColumn.name);

        const paramColumns: ParameterColumn[] = [];
        const argColumns: ArgumentColumn[] = [];

        // FIXME: should not be `appending columns`

        for (let i = 0; i < paramNames.length; i++) {
            const col = new ParameterColumn(paramNames[i], i);
            paramColumns.push(col);
            table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, UnitType.ORIGIN));
            table.setHeader(i + 1, col.name);
        }

        for (let i = 0; i < argLen; i++) {
            const col = new ArgumentColumn(i);
            argColumns.push(col);
            table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, UnitType.ORIGIN));
            table.setHeader(i + paramColumns.length + 1, col.name);
        }

        // setup indicator columns

        const start = table.columnCount;

        for (let i = 0; i < indicators.length; i++) {
            const indicator = indicators[i];
            table.appendColumn(createColumnInfo(ColumnAlign.RIGHT, indicator.unitType));
            table.setHeader(start + i, indicator.indicatorName);
        }

        // setup statistics

        const helpers = [SummaryTable._fnColumn, ...paramColumns, ...argColumns, ...indicators];

        for (let groupId = 0; groupId < statisticGroups.length; groupId++) {
            const group = statisticGroups[groupId];
            for (let i = 0; i < group.length; i++) {
                for (let j = 0; j < helpers.length; j++) {
                    table.setCell([groupId, i], j, helpers[j].getValue(group[i]));
                }
            }
        }

        table.chooseTimeUnit(table.findColumn(StatisticIndicator.Mean.indicatorName));

        // generate report

        this._report = {
            description: !this._descriptionFlag ? undefined : this._generateDescription(indicators, table.timeUnit),
            runtime: !this._runtimeFlag ? undefined : this._generateRuntimeInfo(),
            table: table.render(),
        };

        return this;
    }
}
