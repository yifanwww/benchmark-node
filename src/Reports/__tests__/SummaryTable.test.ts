import { mockConsoleLogger } from '../../__tests__/ConsoleLoggerTestTool';
import { constantsTestTool } from '../../__tests__/constantsTestTool';
import { Column } from '../../Columns';
import { Statistics } from '../../Data';
import { SummaryTable } from '../SummaryTable';

describe(`Test class \`${SummaryTable.name}\``, () => {
    beforeEach(() => mockConsoleLogger());

    it('renders a simple table 1', () => {
        const table = new SummaryTable({ argLen: 0, paramNames: [] });

        table.addStatisticColumns([Column.Mean, Column.Error, Column.StdDev]);
        table.addStats([new Statistics('fn 1', constantsTestTool.measurements.slice(), constantsTestTool.ops, [])]);
        table.addStats([new Statistics('fn 2', constantsTestTool.measurements.slice(), constantsTestTool.ops, [])]);

        table.draw();
        table.drawDescription();
    });

    it('renders a simple table 2', () => {
        const table = new SummaryTable({ argLen: 0, paramNames: [] });

        table.addStatisticColumns([Column.Mean, Column.Error, Column.StdDev]);
        table.addStats([
            new Statistics('fn 1', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
            new Statistics('fn 2', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
        ]);

        table.draw();
        table.drawDescription();
    });

    it('renders a complex table', () => {
        const table = new SummaryTable({ argLen: 2, paramNames: [] });

        table.addStatisticColumns([Column.Mean, Column.Error, Column.StdDev]);
        table.addStats([
            new Statistics('fn 1', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
            new Statistics('fn 2', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
        ]);
        table.addStats([
            new Statistics('fn 1', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
            new Statistics('fn 2', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
        ]);
        table.addStats([
            new Statistics('fn 1', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
            new Statistics('fn 2', constantsTestTool.measurements.slice(), constantsTestTool.ops, []),
        ]);

        table.draw();
        table.drawDescription();
    });
});
