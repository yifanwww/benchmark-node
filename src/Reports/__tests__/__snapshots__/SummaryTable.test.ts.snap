// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test class \`SummaryTable\` renders a complex table 1`] = `
"[96m| Function | arg 0 | arg 1 |      Mean |     Error |    StdDev |[39m
[96m|---------:|------:|------:|----------:|----------:|----------:|[39m
[96m|     fn 1 |     ? |     ? | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
[96m|     fn 2 |     ? |     ? | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
[96m|          |       |       |           |           |           |[39m
[96m|     fn 1 |     ? |     ? | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
[96m|     fn 2 |     ? |     ? | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
[96m|          |       |       |           |           |           |[39m
[96m|     fn 1 |     ? |     ? | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
[96m|     fn 2 |     ? |     ? | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
"
`;

exports[`Test class \`SummaryTable\` renders a complex table 2`] = `
"
"
`;

exports[`Test class \`SummaryTable\` renders a simple table 1 1`] = `
"[96m| Function |      Mean |     Error |    StdDev |[39m
[96m|---------:|----------:|----------:|----------:|[39m
[96m|     fn 1 | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
[96m|     fn 2 | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
"
`;

exports[`Test class \`SummaryTable\` renders a simple table 1 2`] = `
"
"
`;

exports[`Test class \`SummaryTable\` renders a simple table 2 1`] = `
"[96m| Function |      Mean |     Error |    StdDev |[39m
[96m|---------:|----------:|----------:|----------:|[39m
[96m|     fn 1 | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
[96m|     fn 2 | 0.2401 ns | 0.0038 ns | 0.0068 ns |[39m
"
`;

exports[`Test class \`SummaryTable\` renders a simple table 2 2`] = `
"
"
`;
